### ISSUE TITLE
RFC: Revamp isolation

### ISSUE THREAD

**Issue by jvanbruegge** [cyclejs/cyclejs#751]
TL;DR: Use well defined keys for different types of isolation, standardize across drivers.

## Motivation

Currently it is not clear what type of isolation is used just by looking at the code. I think it would help to introduce explicit wrapper functions using Symbols to create a registry of known isolations. This also prevents accidental errors like forgetting the dot when you wanted to do sibling isolation with the DOM driver.

Also the driver can throw an error if it does not support a certain kind of isolation instead of implicitly using a different one.

## Usage

```js
import isolate, { siblingIsolation, defaultIsolation } from '@cycle/isolate';

import Child from './childComponent';

function main(sources) {
    const childSinks = isolate(Child, {
        DOM: siblingIsolation('myKey'),
        '*': defaultIsolation('myKey')
    })(sources);

    return childSinks;
}
```

## Usage in drivers

```js
import { siblingIsolationType, defaultIsolationType, isolationType } from '@cycle/isolate';

class MySource {
    public isolateSink(sink$: Stream<any>, scope: isolationType): Stream<any> {
        if(scope === siblingIsolationType || scope === defaultIsolationType) {
            //do stuff
        } else {
            throw new Error('This driver only supports sibling and default isolation');
        }
    }
    //...
}
```
For non-standard, driver-specific isolations (like the lens isolation with onionify)
```js
import { defaultIsolationType, totalIsolationType, isolationType } from '@cycle/isolate';

export const lensIsolationType = 'LENS_ISOLATION';
export funtion lensIsolation<A, B>(lens: Lens<A, B>) {
    return {
        isolationType: lensIsolationType,
        payload: lens
    };
}

class MySource {
    public isolateSink(sink$: Stream<any>, scope: isolationType | lensIsolationType): Stream<any> {
        if(scope.isolationType === totalIsolationType || scope.isolationType === defaultIsolationType) {
            //do stuff
        } else if(scope.isolationType === lensIsolationType) {
            //do other stuff
        } else {
            throw new Error('This driver only supports total, lens and default isolation');
        }
    }
    //...
}

## Implementation

```js
export const defaultIsolationType = 'DEFAULT_ISOLATION';
export const totalIsolationType = 'TOTAL_ISOLATION';
export const siblingIsolationType = 'SIBLING_ISOLATION';

//see lens isolation for factory functions
```

## Other thoughts

- Instead of keys you could also use ES6 symbols, but I dont see the benefit. With strings we might be able to improve typings with discriminated unions, which is not possible with symbols
- Isolation should not be noticeable from an outside perspective, e.g. currently the DOM driver adds CSS classes on sibling isolation, this should not be the case
- I will open a PR with an example implementation soon

**Comment by mightyiam** [comment]
Sounds beautiful.

**Comment by jvanbruegge** [comment]
cc @staltz @Widdershin

**Comment by staltz** [comment]
Hi Jan, thanks for the RFC. This is like the elephant in the room, because the isolation API is not the best.

Before commenting, here are some important references to discussions that we had about this in the past:

- https://github.com/cyclejs/cyclejs/issues/526
- https://github.com/cyclejs/cyclejs/issues/369
- https://github.com/cyclejs/cyclejs/issues/296
- https://github.com/cyclejs/cyclejs/issues/167

Specially #526 was an important discussion.

## History of Isolation

And for reference, here's the "History of Isolation" in a nutshell:

- It used to be hard coded as "namespaces"
- "Scopes" was suggested https://github.com/cyclejs/cyclejs/issues/167
- cycle/isolate was born, there was only strings as scopes
- onionify introduced lens object as scope (the first non-string type of scope)
- isolate was generalized to be flexible, introduced also `'*'` and per-channel scopes

Given that history, we ended up with an API that is not explicit. It is somewhat magic. So I strongly agree with this argument as a starting point:

> Currently it is not clear what type of isolation is used just by looking at the code.

And we should look for an API that makes it clear what type of isolation is being used. That said, I think if we are going to make a breaking change, we should really step back and reconsider the whole thing, and redesign the API for scratch now that we have learned a lot about isolation.

## Requirements that we have learned

So I'm going to try to layout some basic requirements for isolation, ignoring syntax:

1. It should be possible to specify which channels are isolated and which are not
2. The type of the scope depends on the channel
3. Some channels have different flavors (sibling, total, etc) of isolation
4. From a beginners perspective, the API should *teach*, not *confuse*
5. From an expert's perspective, the API should not be tedious to write

I also thought about whether random scopes are an important requirement, but I didn't feel convinced that they are important. In fact, they just add impurity, and make the code just a little bit shorter. I also considered whether it is an important requirement to be able to isolate all of the channels at once with the same shared scope (`isolate(comp, scope)`), and I noticed it is a nice-to-have feature but not so important.

## Proposal A

So let's look at one proposed syntax:

```js
    const childSinks = isolate(Child, {
        DOM: siblingIsolation('myKey'),
        '*': defaultIsolation('myKey')
    })(sources);
```

And how it weighs against the requirements:

1. Yes we can specify which channels
2. A bit weird because `siblingIsolation` comes from `@cycle/isolate` and as a function it returns a certain type which may or may not be suitable for a specific channel. So I mean that anything that produces a scope should be coming from the channel (driver package), not from somewhere global. If it comes from somewhere global then it means that all of the channels have to support string scopes, which may be okay, but supporting string scopes everywhere is not strong requirement.
3. Also weird because the isolation flavors come from `@cycle/isolate`, so drivers would have to throw errors for those flavors that are not supported. If flavors are imported for each channel, we wouldn't have this problem.
4. Yes I think this API teaches beginners
5. Maybe experts will be okay with this API, but it's almost becoming tedious already

## Proposal B

Here's another strawman syntax proposal that I came up just by imagining "what would be the perfect API for this?"

```js
import {isolateDOMFromSiblings} from '@cycle/dom';
import {isolateHTTP} from '@cycle/http';

const childSinks = Child
  |> isolateDOMFromSiblings('foo')
  |> isolateHTTP('foo')
  |> sources
```

Notes:
- `isolateDOMFromSiblings :: string -> Component -> Component`
- `isolateHTTP :: string -> Component -> Component`
- Name "isolateDOMFromSiblings" can be bikesheded
- There would also be `import {isolateDOMFromAll} from '@cycle/dom'`
- Of course we can't count on `|>` in ESNext, but let's imagine we live in a perfect world
- There would be no more `@cycle/isolate`

And how it weighs against requirements:

1. Yes, with this chained API, each time we call `isolateSomething` we get a new component
2. Yes, type of the scope is determined by the package from the imports
3. Each package exports its own flavors of isolation, so there is no need for some drivers to throw on unsupported flavors.
4. I hope from a beginners perspective, this chained API looks simple: we take the `Child` component, then isolate the DOM channel from its siblings, then we isolate the HTTP channel, then we give it some source streams.
5. Chained APIs are nice to use from an experts perspective, I hope

- - -

Note: everything I said above is debatable. :)

**Comment by jvanbruegge** [comment]
>  So I mean that anything that produces a scope should be coming from the channel 

That was to standardize common isolation schemes to prevent each driver to name the same concept different / provide different semantics

> If it comes from somewhere global then it means that all of the channels have to support string scopes

No, the only requirement for a driver/utility would be to implement defaultIsolation. The payload of which can be anything (the Type would be generic)

> Also weird because the isolation flavors come from @cycle/isolate, so drivers would have to throw errors for those flavors that are not supported. If flavors are imported for each channel, we wouldn't have this problem.

A good argument, but it clashes with standardization. Do we want to provide familiar experiences, or do we want each driver to provide its own solutions?

-----------------------------------------
About proposal B: First of all I like it.

> Of course we can't count on |> in ESNext, but let's imagine we live in a perfect world

In my opionion |> is far from perfect, but lets look at it without magic syntax sugar, ideally the `isolate*` functions take a component and return a new one, so you can do this:
```js
R.compose(
    isolateDOMFromSibling('foo'),
    isolateHTTP('foo'),
    Child
)(sources);
```

**Comment by jvanbruegge** [comment]
If you look closer, both proposals are almost identical:
```js
isolate(Child, {
    DOM: siblingIsolation('foo'),
    HTTP: defaultIsolation('foo')
})(sources);

compose(
    isolateDOMFromSibling('foo'),
    isolateHTTP('foo'),
    Child
)(sources);
```
This is basicly just a style difference.

The main difference is if you want standardization of isolation schemes or not. With isolation you can to this:
```js
isolate(Child, defaultIsolation('foo'))(sources)
```
WIthout you would have to have a helper function (which may be good, because it is more explicit)
```ts
function defaultIsolation(exeptions, key, Child) {
    const isolations = {
        DOM: isolateDOMTotally(key),
        HTTP: isolateHTTP(key),
        ...exceptions
     };
    return R.compose(...Object.keys(isolations).map(k => isolations[k]), Child);
}
defaultIsolation({}, 'foo', Child)(sources);
```
But this would again implictly require the drivers to implement string key isolation.

I think it is good that every driver shoud implement a string based isolation. Most of the time you just want a simple total isolation.

**Comment by jvanbruegge** [comment]
So to summarize my opinion: I dont care how the API looks (propsal A or B), but I am strongly in favor of standardization and a default isolation that every driver implements

**Comment by abaco** [comment]
For me the question is: which one is easier to understand, a standardized API or an explicit API? I agree with @janvanbruegge that in this case a standard set of concepts would make things simpler.

If we are going the standardized way I would add:

- an identity element for isolation, i.e. `noneIsolationType`
- every channel must accept string scopes, so we can still do `isolate(Child, 'myKey')(sources)`
- every channel must accept *default*, *total*, *sibling*, and *none* isolation scopes (it might not be a common need, but total isolation of the `onion` or `HTTP` channels does make sense)

**Comment by jvanbruegge** [comment]
@abaco 
> an identity element for isolation, i.e. noneIsolationType

I would just leave that out of the `isolate` call. As implicit scope generation will be (should be) removed, leaving out a key would mean no isolation. Less to type, same sematics.

> every channel must accept string scopes, so we can still do isolate(Child, 'myKey')(sources)

I would say a driver should only accept `Scope` types, so this would look like this: `isolate(Child, defaultIsolation('myKey'))(sources)`

> every channel must accept default, total, sibling, and none isolation scopes (it might not be a common need, but total isolation of the onion or HTTP channels does make sense)

I dont think it makes sense for every driver to have those. I think only one string scoped isolation should be required (which is also the default then). But I would love to hear why you want to force every driver to have them.

**Comment by whitecolor** [comment]
```ts
R.compose(
    isolateDOMFromSibling('foo'),
    isolateHTTP('foo'),
    Child
)(sources);
```
Should probably be:

```ts
R.compose(
    isolateDOMFromSibling('foo'),
    isolateHTTP('foo'),
)(Child)(sources);
```

@jvanbruegge 

> but I am strongly in favor of standardization and a default isolation that every driver implements

I think that different drivers tend to provide different kinds and experiences of isolations, and some may not need it at all.  There is not only visual/aesthetic differences between **A** and **B** options.

So, my current take on this issue is to agree with @staltz **Proposal B**: to get rid of current generic `isolate` module in favor of driver specific isolation helpers:

```ts
import {isolateEffect} from 'effect-driver'
....
// 'sourceKey' and 'sinkKey' would have default agreed values (like `DOM`, 'HTTP'), 
// but could be altered
isolateEffect('scopeValue', 'sourceKey', 'sinkKey')(Comp)
```
Actually, I see mostly benefits of this move:

1) It provides more functional, transparent and decoupled approach (driver and apps are not tired with `isolate` module any more), so we just get rid of excessive coupling.
Isolation can be **standardized** by providing generic rules and even utility helper to build driver specific isolation functions that follow standard approach.
2) Full isolation documentations will be available in the code, currently, because you just have generic `isolate` module we don't have TS docs on isolation of particular module besides one on the web.
3) Good (at least better) TS typing can be done more easily (notice, current `isolate` doesn't type at all - returns `any`), mostly driver/modules isolation functions doesn't change typings at all, but I think even typings of `onionify` could be done with success. 

The only problem with API for application of multiple isolate functions to the component (like `R.compose` and `R.pipe`) TS has current problems with this case.

**Comment by staltz** [comment]
Jan was right about the syntax, that's the least important part. The real issue is from where will the scopes come from: `@cycle/isolate` or from each driver package.

Standardization is usually about providing a common interface, or a common API, or a common method, or just us achieving consensus on **how** to perform something.

In the case of isolation, we are actually talking about different semantics, not different methods. This is why I think standardization doesn't apply. Let me explain how.

"Sibling DOM isolation" **means** something different to "Sibling HTTP isolation", even though both use the word sibling. In fact each channel or driver needs to define what isolation means for that channel. And effect we currently do this in our documentation: https://cycle.js.org/api/dom.html#isolation-semantics . So it seems suitable that each driver would *provide* functions that come with that semantic *built-in*. The **standard** in this case would be the function signature, which we would like all drivers to have a consensus. Note that again standard he refers to "how", not "what".

To illustrate what could happen if we put generic semantics default/total/sibling from a common package, we would end up with "sub-semantics". Total isolation in the DOM would works slightly different to total isolation in onionify or HTTP or something else, and and then to mitigate that confusion we would have to introduce yet more "generic semantics": `import {totalIsolationTypeA} from '@cycle/isolate'` and `import {totalIsolationTypeB} from '@cycle/isolate'`.

And as a different example from another context, think of `Functor` from functional programming languages. It just defines an interface, function signature, and some basic mathematical properties that should apply. But it doesn't define semantics. In the case of list, the Functor function will "map" each element to another element. But in the case of the IO Monad, the Functor function will simply apply an argument to the element wrapped by the monad.

**Comment by staltz** [comment]
Adding to that last example, if we were to model this in Haskell style, there would be the "Isolatable" typeclass with the generic function `isolate :: scope -> Component -> Component`. In fact we could standardize that all drivers export `isolate` so you could do `import {isolate} from '@cycle/dom'`. In fact, each driver already "has an instance of that typeclass" because it implements `isolateSource` and `isolateSink`.

That would be nice, but the problem is that each driver may have more than one `isolate`. In the DOM driver we already have two "instances of the isolateSource typeclass": https://github.com/cyclejs/cyclejs/blob/5be4faa0eac6f93470a25281bd6167a61bb0de93/dom/src/isolate.ts#L6-L10

This is why we need different names coming from the drivers. It wouldn't be a big change because we already have each driver implementing and exporting its own `isolateSource`/`isolateSink` instances.

**Comment by jvanbruegge** [comment]
> "Sibling DOM isolation" means something different to "Sibling HTTP isolation", 

I think this is where we should start. Even if they are in seperate modules, they will be still called the same. This would confusing for a newcomer.

As you said a typeclass is a signature + optional mathematical rules. If you standardize the meaning of sibling and total this would fulfil this.

In my opinion, we should just define those two terms:
Isolation type | events in the parent         | events in the child
-----------------|----------------------------------|---------------------------------
total               | only accessible by parent | only accessible from child
sibiling           | only accessible by parent | accessible by parent and child

The problem is, should we force any of those to be available in the drivers? Because for some this might not make sense. E.g. total isolation with onionify, because the parent has access to the child state. But on the other hand it might be useful to add total isolation to onionify.

TL;DR: 
- Regardless if isolations come from one module or from each driver, terms like `total` and `sibling` should just have one meaning that is consistent across the codebase.
- It might be useful to force all drivers to implement sibling and total isolation schemes, but that might not be possible

**Comment by staltz** [comment]
@jvanbruegge Notice that when you standardize the meaning of "total" to be "only accessible from child", we are just pushing the problem of meaning elsewhere: now we need to define the **meaning** of the word "accessible".

**Comment by staltz** [comment]
Also, an important note is that the "total or siblings" distinction currently **only** exists for the DOM driver. Even though the HTTP driver docs mention "sibling isolation", it just means "normal isolation". So I find it complicated to extract total-versus-siblings from the DOM driver and put it everywhere, when in fact it only makes sense for the DOM, so far.

**Comment by jvanbruegge** [comment]
well it would make sense for the HTTP driver too. Total isolation for HTTP means only the child can receive the responses of its own requests. Sibling isolation means that requests and responses can be accessed in the parent.
I'm not sure if total isolation would be helpful with HTTP, because it would prevent the usage of a wrapper to augment the requests with e.g. auth data

**Comment by jvanbruegge** [comment]
But yeah, after thinking about it for a longer time, I think it would be good to let each driver define what isolations are helpful. We should just watch out on the implementation/documentation side, that we do not use the same terms for different meanings.

**Comment by staltz** [comment]
And this is a topic we need to give a *lot* of careful thought and API design. First we made cycle/isolate, and it's okay that it has flaws. But Isolation 2.0 should do it right, if we're going to ask people to migrate a breaking change.

**Comment by jvanbruegge** [comment]
We should also think about pull a bit, just to make sure this stays somewhat forward compatible in the future

**Comment by abaco** [comment]
I recognize the advantages of something like @staltz's proposal B, but let me explore the original proposal a bit more.

> Notice that when you standardize the meaning of "total" to be "only accessible from child", we are just pushing the problem of meaning elsewhere: now we need to define the meaning of the word "accessible".

The meaning of "accessible" can be standardized if all drivers have `source.select(scope)`. For example in the following code, because sibling isolation is used, `childXYZSource` would be identical to the XYZ source in the child component.

```typescript
const childSinks = isolate(Child, siblingIsolation('myKey'))(sources);
const childXYZSource = sources.XYZ.select('myKey');
```

In case of total isolation `select('myKey')` would instead return an empty source. This seems a perfectly clear requirement that could be enforced in every driver.

I also proposed that all drivers implement both sibling and total isolation just because it's less surprising than throwing errors. It might be useless, but you can give a sense to a totally isolated `onion` source: the parent can't access the child's state. Computing `3 + 0` is also useless, but it makes perfectly sense. That's also why I thought a `noneIsolation` wouldn't hurt.

I like that proposal B is explicit, and I appreciate @whitecolor's motivations. My biggest objection now is that it's quite verbose, so I'm afraid I'll reinvent a generic `isolate` in my projects...

**Comment by whitecolor** [comment]
@abaco I agree that current API is just simple and convenient to use in the simplest cases (which are mostly the case for me), though only with typings saved =).

> I also proposed that all drivers implement both sibling and total isolation just because it's less surprising than throwing errors. 

How and what would you want to isolate in fully asymmetrical driver ala web-socket, that sends messages and receives messages that are not coupled on driver's level. Or I wonder what is "the case" for total/siblings isolation of state (onionify)? It would just break consistency and make things obscure.

**Comment by abaco** [comment]
@whitecolor You're right, I wasn't thinking of asymmetrical drivers like websockets. In such cases isolation would have to do nothing - quite confusing, I agree! As for total isolation in onionify, it would provide one more way (there are others) to keep a component's state local. I see that these aren't real advantages. I liked the idea of a standard isolation vocabulary, but it turns out it would hide a lot of weirdness instead of conveying neat concepts. The only issue it would address is verbosity, but for that it would suffice that each driver exposed a default isolation function.

If **Proposal B** gets finally implemented (at this point I endorse it) I think I would use the following pattern in my code:

```typescript
export function button(sources: Sources): Sinks { ... }

export function makeIsolatedButton(scopeName: String, stateLens: Lens): Button {
  return R.compose(
    isolateDOMFromSibling(scopeName),
    isolateStateWithLens(stateLens),
    isolateHTTP(scopeName)
  )(button);
}
```

Each component file would export the bare component, but also a factory that create already isolated components (or several factories with different isolation strategies). IMO this pattern provides a good balance of explicitness and conciseness.

**Comment by whitecolor** [comment]
> As for total isolation in onionify, it would provide one more way (there are others) to keep a component's state local.

Onionify already provides this kind of "isolation", which made "isolated" state local to a component. But it is not a total isolation, this means that changes applied to the (root) state that by "isolated" component can be observed by others, the first: you will hardly be able to implement the opposite, the second: you definitely don't want to do this, case it would really break app consistency in terms of handling state (value over time). 

When dealing with DOM it is a little bit different thing, that is why total isolation can be useful there, because it can be not obvious how to restrict parent components from reactions to isolated child events, so you may get totally isolated component in terms of DOM events, it probably comply with the concept of web components, shadow dom, etc.

> IMO this pattern provides a good balance of explicitness and conciseness.

If to ignore clumsy typings of `R.compose` (which is TS current problem), but in form of common factory, it is ok.

**Comment by staltz** [comment]
About total isolation for onionify, let me see if I understand: it would probably be implement with a `parent._secret`-style of nesting, right? Because in the end of the day, due to how onionify works, the way a parent passes data down to a child is through the state object. The whole idea is that there is only one state object in the whole application, not many.

Unless you mean total isolation of `child` would be similar to calling `onionify(child)`, in which case we don't need to implement anything new, because `onionify == totalIsolation` 😎

**Comment by abaco** [comment]
I was thinking of something similar to calling `onionify(child)`, only through `isolate`. However at this point I consider the whole thing off-topic... :)

Back to proposal B... what about channel names? Should each `isolateXXX` also take a channel name in case the default one isn't used? This way:

```typescript
isolateHTTP('foo')         // isolates HTTP channel named "HTTP"
isolateHTTP('foo', 'AJAX') // isolates HTTP channel named "AJAX"
```
Or maybe each `isolateXXX` works on a single channel, and a utility function lifts it to a component? So:
``` typescript
chan('DOM', isolateHTTP)('foo')
chan('AJAX', isolateHTTP)('foo')
```
Or perhaps a single function that combines several *channel utilities* to create a *component utility*:
```typescript
const isolate = componentUtility({
    DOM: isolateDOMFromSibling('foo'),
    onion: isolateStateWithLens(stateLens),
    HTTP: isolateHTTP('foo'),
});
const buttonSinks = isolate(button)(sources);
```

**Comment by jvanbruegge** [comment]
@abaco isn't the channel the `foo`?

**Comment by whitecolor** [comment]
> Should each isolateXXX also take a channel name in case the default one isn't used?

Of course, that could be implemented (and even typed well) without problems I pointed this [in my comment](https://github.com/cyclejs/cyclejs/issues/751#issuecomment-347942241)

> Thinking of how to implement the latest two, I realized it's actually impossible, 

Of course, it is possible, everything is possible in JavaScript =)

**Comment by staltz** [comment]
Just had a random idea to use template strings:

```js
R.compose(
    isolateDOMFromSibling`foo`,
    isolateHTTP`foo`,
)(Child)(sources);
```

Not sure if it helps much, but I just wanted to record the idea here for discussion.

**Comment by aronallen** [comment]
Just some input here, I might be a bit late to the game.
We are running Cycle in WebWorkers and sometimes we need to perform single mutations on the DOM that don't mandate a full patch, alternatively we might want to read a single property on a node without any DOM event happening on that node.
Think of this as a jQuery command we declare on a single selection.
For this we have created a DOMEffect driver, that shares it's isolation table with the regular DOM driver.
The idea is that we can declare an effect like this.

```
just(effect('input', 'value', ''))
```
The above example would set the value of all inputs to empty string.

```
DOMEffect.select('input', { value: String()}, periodic(1000)).map(console.log)
```
The above would return the value of the input every second.

Some of our constraints are very specific, because we are running in WebWorkers, but the idea that drivers could share isolation tables is quite useful.

To achieve the above we need the DOMDriver to return its isolation table, so it can be shared with the DOMEffect driver. We have worked around this by making our own DOMDriver.

Potentially drivers could return an optional IsolationTable symbol, that references its internal isolation table, and that we somehow can refer to this from other drivers. There is a problem here of order, DOM must be declared before DOMEffect, but I am sure we could find a great solution.
